Optimize some per-document hot paths in the geoip processor

joegallo opened the pull request:
Each commit is a different optimization, and they all matter (just a little). Here's the flamegraph before (with areas highlighted):

![Screenshot 2025-01-24 at 12 26 36 PM](https://github.com/user-attachments/assets/e2a4049a-36c3-48be-9e01-074b75f1cfbb)

And after:

![Screenshot 2025-01-24 at 12 26 44 PM](https://github.com/user-attachments/assets/447eb810-a652-46fb-aed1-0c7c09b347f5)

joegallo committed:
Cache this toString

joegallo committed:
Switch to millisecond accuracy

This saves us the expensive time to retrieve and set the nanoseconds
part of the Instant.

joegallo committed:
Avoid unlikely and more-expensive settings call path

joegallo committed:
Update docs/changelog/120824.yaml

masseyke commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java:
```diff
@@ -206,12 +206,27 @@ public static Metadata fromXContent(XContentParser parser) {
 
         public boolean isCloseToExpiration() {
-            return Instant.ofEpochMilli(lastCheck).isBefore(Instant.now().minus(25, ChronoUnit.DAYS));
+            final Instant now = Instant.ofEpochMilli(System.currentTimeMillis()); // millisecond precision is sufficient (and faster)
```
I think we could pass in a ThreadPool, and use ThreadPool::absoluteTimeInMillis, which would be significantly faster since it's cached. Same goes in the other place below.

joegallo replied:
That makes sense. I'll see if I can make that happen.

joegallo replied:
I can't make it pass tests! I'm going to open a new ticket to track it.

masseyke commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java:
```diff
@@ -206,12 +206,27 @@ public static Metadata fromXContent(XContentParser parser) {
+            // to show up in a flame graph.
+            final long valid;
+            if (settings.hasValue("ingest.geoip.database_validity")) {
+                valid = settings.getAsTime("ingest.geoip.database_validity", THIRTY_DAYS).millis();
```
Is this slow only when there is no value (b/c we generate an exception or something)? Or is it always slow? If the latter, should we calculate this outside of this method, and have a settings listener listening for updates to the setting?

joegallo replied:
I added another comment here in the code -- the setting itself is kindof a hack to get around the situation where we want to be able to test this code but we can't make it possible to actually change the value of the setting in 'the real world'.

I'm not at all opposed to continuing to grind on this, but I'd love to do that in a different PR.

joegallo replied:
> If the latter, should we calculate this outside of this method, and have a settings listener listening for updates to the setting?

I took a closer look at the idea of the settings listener, and I think the reason we're not doing any of that right now is specifically because the setting in question isn't a 'real setting' -- AFAICT all the settings update listener type methods require a proper `Setting` object to hang off of, which we don't have in this case. 😬

Somebody could probably get pretty far by extracting our current code and **manually** watching for settings changes like this:

```
+        clusterService.addListener(event -> {
+            if (event.metadataChanged() && event.state().metadata().settings() != event.previousState().metadata().settings()) {
+                logger.info("metadata changed and settings changed");
+                [... process the settings update here, check if the database validity has changed, etc ...]
+            }
+        });
```

joegallo committed:
Add an explantory comment

joegallo committed:
Use a cached task state rather than rebuilding it

joegallo commented:
I need to snag a profile, but a6a41355728db717a34f6d173a64edccc9a9c6a9 is a very nice change here.

joegallo committed:
Merge branch 'main' into ingest-geoip-microoptimizations

joegallo committed:
Rewrite this test assertion

Make it a little more obvious in the test failure message that
GeoLite2-ASN.mmdb isn't particularly special.

joegallo committed:
Merge branch 'main' into ingest-geoip-microoptimizations

joegallo committed:
Revert "Use a cached task state rather than rebuilding it"

This reverts commit a6a41355728db717a34f6d173a64edccc9a9c6a9.

joegallo commented:
I kicked some changes out of this PR, and I'm punting some of what I would still like to see done onto https://github.com/elastic/elasticsearch/issues/121208. We can come back to this in the future and do even better.

masseyke reviewed (approved):
LGTM

joegallo committed:
Add a missing import

joegallo committed:
Merge branch 'main' into ingest-geoip-microoptimizations

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/121226)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/121226">](https://github.com/elastic/elasticsearch/pull/121226) |
